unixfs/src/pb.rs

```diff
@@ -0,0 +1,233 @@
+
+    fn try_from(data: Option<&'a [u8]>) -> Result<Self, Self::Error> {
+        use quick_protobuf::{BytesReader, MessageRead};
+
+        let data: &'a [u8] = data.ok_or(UnixFsReadFailed::NoData)?;
+        let mut reader = BytesReader::from_bytes(data);
+        UnixFs::from_reader(&mut reader, data).map_err(UnixFsReadFailed::InvalidUnixFs)
+    }
+}
+
+// These should be derived by the pb-rs
+impl<'a> TryFrom<&'a [u8]> for merkledag::PBNode<'a> {
+    type Error = quick_protobuf::Error;
+
+    fn try_from(data: &'a [u8]) -> Result<Self, Self::Error> {
+        use quick_protobuf::{BytesReader, MessageRead};
+        merkledag::PBNode::from_reader(&mut BytesReader::from_bytes(data), data)
+    }
+}
+
+/// Intermediate conversion structure suitable for creating multiple kinds of readers.
+pub(crate) struct FlatUnixFs<'a> {
+    pub(crate) links: Vec<PBLink<'a>>,
+    pub(crate) data: UnixFs<'a>,
+}
+
+impl<'a> TryFrom<&'a [u8]> for FlatUnixFs<'a> {
+    type Error = UnixFsReadFailed;
+
+    fn try_from(data: &'a [u8]) -> Result<Self, Self::Error> {
+        let PBNode {
+            Links: links,
+            Data: data,
+        } = merkledag::PBNode::try_from(data).map_err(UnixFsReadFailed::InvalidDagPb)?;
+
+        let inner = UnixFs::try_from(match data {
+            Some(Cow::Borrowed(bytes)) if !bytes.is_empty() => Some(bytes),
+            Some(Cow::Owned(_)) => unreachable!(),
+            Some(Cow::Borrowed(_)) | None => return Err(UnixFsReadFailed::NoData),
+        })?;
+
+        Ok(FlatUnixFs { links, data: inner })
+    }
+}
+
+#[cfg(test)]
+impl<'a> FlatUnixFs<'a> {
+    pub fn range_links(&'a self) -> impl Iterator<Item = (PBLink<'a>, Range<u64>)> {
+        assert_eq!(self.links.len(), self.data.blocksizes.len());
+
+        let zipped = self
+            .links
+            .clone()
+            .into_iter()
+            .zip(self.data.blocksizes.iter().copied());
+
+        // important: we have validated links.len() == blocksizes.len()
+        RangeLinks::from_links_and_blocksizes(zipped, Some(0))
+    }
+}
+
+pub(crate) struct RangeLinks<I> {
+    inner: I,
+    base: u64,
+}
+
+impl<'a, I> RangeLinks<I>
+where
+    I: Iterator<Item = (PBLink<'a>, u64)>,
+{
+    /// `start_offset` is the offset of the current tree when walking the graph.
+    pub fn from_links_and_blocksizes(zipped: I, start_offset: Option<u64>) -> RangeLinks<I> {
+        RangeLinks {
+            inner: zipped,
+            base: start_offset.unwrap_or(0),
+        }
+    }
+}
+
+impl<'a, I> Iterator for RangeLinks<I>
+where
+    I: Iterator<Item = (PBLink<'a>, u64)>,
+{
+    type Item = (PBLink<'a>, Range<u64>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.inner.next() {
+            Some((link, blocksize)) => {
+                let returned_base = self.base;
+                self.base += blocksize;
+                Some((link, returned_base..(returned_base + blocksize)))
+            }
+            None => None,
```

Comment:
nit: I think `self.inner.next().map(|(link, blocksize)| ...)` would work nicely here

Reply:
I guess in general I've learned to be cautious around borrowing &mut self into closures but it might work here.

Reply:
Can be changed, did change.